build-tools/packages/build-cli/src/library/commands/generateSourceEntrypoints.ts

```diff
@@ -0,0 +1,299 @@
+
+		const { outFileSuffix, outFileAlpha, outFileBeta, outFileLegacy, outFilePublic, outDir } =
+			optionDefaults;
+
+		const mapSrcQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined> = new Map([
+			[`${outDir}${outFileAlpha}${outFileSuffix}`, ApiTag.alpha],
+			[`${outDir}${outFileBeta}${outFileSuffix}`, ApiTag.beta],
+			[`${outDir}${outFilePublic}${outFileSuffix}`, ApiTag.public],
+			[`${outDir}${outFileLegacy}${outFileSuffix}`, ApiTag.legacy],
+		]);
+
+		const { mapSrcApiTagLevelToOutput } = getOutputConfiguration(
+			packageJson,
+			tsConfig,
+			mapSrcQueryPathToApiTagLevel,
+			this.logger,
+		);
+
+		if (mapSrcApiTagLevelToOutput.size === 0) {
+			throw new Error(
+				`There are no package exports matching requested output entrypoints:\n\t${[
+					...mapSrcQueryPathToApiTagLevel.keys(),
+				].join("\n\t")}`,
+			);
+		}
+
+		const promises: Promise<void>[] = [];
+		promises.push(
+			generateSourceEntrypoints(mainEntrypoint, mapSrcApiTagLevelToOutput, this.logger),
+		);
+
+		// All of the output actions (deletes of stale files or writing of new/updated files)
+		// are all independent and can be done in parallel.
+		await Promise.all(promises);
+	}
+}
+
+// Reads and parses the `package.json` file in the current directory.
+async function readPackageJson(): Promise<PackageJson> {
+	const packageJson = await fs.readFile("./package.json", { encoding: "utf8" });
+	return JSON.parse(packageJson) as PackageJson;
+}
+
+// Reads and parses the `tsconfig.json` file in the current directory.
+async function readTsConfig(): Promise<TsConfigJson> {
+	const tsConfigContent = await fs.readFile("./tsconfig.json", { encoding: "utf8" });
+	// Trim content to avoid unexpected whitespace issues
+	const trimmedContent = tsConfigContent.trim();
+
+	// Remove trailing commas before parsing
+	const sanitizedContent = trimmedContent.replace(/,\s*([\]}])/g, "$1");
+
+	// Parse and validate JSON content
+	return JSON.parse(sanitizedContent) as TsConfigJson;
+}
+
+/**
+ * Generates mappings based on the provided `package.json` and `tsconfig.json`. It establishes a relationship between source paths
+ * and API tag levels, and maps these tags to their corresponding export data.
+ *
+ * @param packageJson - `package.json` content.
+ * @param tsconfig - `tsconfig.json` content.
+ * @param mapSrcQueryPathToApiTagLevel - Maps source query paths or regular expressions to their corresponding API tags.
+ * @param logger - An optional logger for logging messages or warnings during processing.
+ * @returns Maps API tags to their associated export data.
+ */
+function getOutputConfiguration(
+	packageJson: PackageJson,
+	tsconfig: TsConfigJson,
+	mapSrcQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined>,
+	logger?: CommandLogger,
+): {
+	mapSrcApiTagLevelToOutput: Map<ApiTag, ExportData>;
+} {
+	let emitDeclarationOnly: boolean = false;
+	if (tsconfig.compilerOptions?.emitDeclarationOnly !== undefined) {
```

Comment:
Why is this needed?

Reply:
If emitDeclarationOnly is set to true in the tsconfig, the types field should be referenced, as the default field may not be present. For example, datastore-definitions only emits type declarations.

https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore-definitions/package.json#L18
https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore-definitions/tsconfig.json#L8

Reply:
I'm confused - what is driving the decision of what files to create and what files to read as input? I thought it was using the index.ts file to decide what the API levels in use are. Is it reading the package.json exports to determine that? Seems like there should be one input - the list of all exports. Why does the types or default field matter?